utils/lib/Build.hs

```diff
@@ -0,0 +1,177 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Build where
+
+import Data.Text
+    ( Text )
+import qualified Data.Text as T
+
+import CommonBuild
+    ( BuildkiteEnv
+    , CoverallsConfig
+    , DryRun
+    , ExitCode (ExitSuccess)
+    , cacheGetStep
+    , cachePutStep
+    , cleanBuildDirectory
+    , cleanupCacheStep
+    , doMaybe
+    , echo
+    , getBuildkiteEnv
+    , getCacheConfig
+    , isBorsBuild
+    , onDefaultBranch
+    , purgeCacheStep
+    , run
+    , setupBuildDirectory
+    , timeout
+    , uploadCoverageStep
+    , void
+    , weederStep
+    , when
+    , whenRun
+    , (.&&.)
+    )
+
+import BuildArgs
+    ( BuildArgs (BuildArgs, command, options)
+    , Command (Build, CleanupCache, PurgeCache)
+    , RebuildOpts (RebuildOpts, optBuildDirectory, optCacheDirectory, optDryRun)
+    , parseArgs
+    )
+
+import Data.Maybe
+    ( fromMaybe )
+import System.Exit
+    ( exitWith )
+
+
+-- | Run the CI build step.
+--
+-- The specific command and its options can be passed as command line
+-- arguments. See module 'BuildArgs'.
+doBuild
+  :: LibraryName
+  -> Optimizations
+  -> (BuildkiteEnv -> QALevel)
+  -- ^ Function used to determine the QA level from the Buildkite environment variables. See
+  -> [TestRun]
+  -> CoverallsConfig
+  -> IO ()
+doBuild (LibraryName whichLibrary) optimizations buildkiteEnvToQALevel testRuns coverallsConfig = do
+  BuildArgs {options, command} <-
+    parseArgs $ "Build " ++ whichLibrary ++ " with stack in Buildkite"
+  let RebuildOpts { optBuildDirectory, optCacheDirectory, optDryRun } = options
+  builkiteEnv <- getBuildkiteEnv
+  let qaLevel = maybe QuickTest buildkiteEnvToQALevel builkiteEnv
+  cacheConfig <- getCacheConfig builkiteEnv optCacheDirectory
+  case command of
+    Build -> do
+      doMaybe (setupBuildDirectory optDryRun) optBuildDirectory
+      whenRun optDryRun $ do
+        cacheGetStep cacheConfig
+        cleanBuildDirectory (fromMaybe "." optBuildDirectory)
+      buildResult <- buildStep optDryRun optimizations qaLevel testRuns
+      when (shouldUploadCoverage qaLevel) $
+        uploadCoverageStep coverallsConfig optDryRun
+      whenRun optDryRun $ cachePutStep cacheConfig
+      void $ weederStep optDryRun
+      exitWith buildResult
+    CleanupCache ->
+      cleanupCacheStep optDryRun cacheConfig optBuildDirectory
+    PurgeCache ->
+      purgeCacheStep optDryRun cacheConfig optBuildDirectory
+
+newtype LibraryName = LibraryName String deriving Show
+
+data Optimizations = Standard | Fast deriving (Show, Eq)
+
+data QALevel = QuickTest | FullTest deriving (Show, Eq)
```

Comment:
I think the `QALevel` is perhaps specific to our own project on `cardano-wallet` because we've categorized tests in two categories:

- unit
- integration

For example:  `cardano-wallet-jormungandr:test:unit` and `cardano-wallet-jormungandr:test:integration`. The latter requires extra setup as it needs to connect to an actual running node. 

I am not sure this peculiarity is common to many projects and should be part of the skeleton code. Could be taken away IMO, or rebranded simply as "skip" options being passed down to stack (the stack `--skip`, not the hspec one).